rllib/evaluation/rollout_worker.py

```diff
@@ -1945,12 +1962,40 @@ def _determine_spaces_for_multi_agent_dict(
         if env_act_space is None:
             env_act_space = spaces.get("__env__", [None, None])[1]
 
-    for pid, policy_spec in multi_agent_dict.copy().items():
+
+    for pid, policy_spec in multi_agent_policies_dict.copy().items():
         if policy_spec.observation_space is None:
             if spaces is not None and pid in spaces:
                 obs_space = spaces[pid][0]
             elif env_obs_space is not None:
-                obs_space = env_obs_space
+                # Multi-agent case AND different agents have different spaces:
+                # Need to reverse map spaces (for the different agents) to certain
+                # policy IDs.
+                if isinstance(env, MultiAgentEnv) and env._spaces_in_preferred_format:
+                    obs_space = None
+                    mapping_fn = policy_config.get("multiagent", {})\
+                        .get("policy_mapping_fn", None)
+                    if mapping_fn:
```

Comment:
just to double check, what does it mean if there is no mapping_fn?
do we assume that all agents get mapped to __default__ policy in that case?
do we need to handle the case?

Reply:
This is mostly for testing puproses, where RolloutWorkers may receive only a partial config dict.
In the normal case, they get handed a complete trainer config dict with all these keys (multiagent, policy_mapping_fn) available.